# Proyek Analisis Data: E-Commerce Public Dataset
- **Nama:** Fathur Rahman Al Farizy
- **Email:** fathur.alfarizy02@gmail.com
- **ID Dicoding:** fralfrzy

## Menentukan Pertanyaan Bisnis

1. Bagaimana perkembangan total penjualan seluruh produk tiap bulannya pada periode 2016-2018?
2. Kategori produk apa yang mendapatkan hasil penjualan tertinggi selama periode 2016-2018?
3. Kategori produk apa yang mendapatkan hasil penjualan terendah selama periode 2016-2018?
4. Kategori produk apa yang mendapatkan rata-rata skor review terendah dari pelanggan pada periode 2016-2018? 
5. Faktor apa yang menyebabkan kategori produk tertentu mendapatkan rata-rata skor review terendah dari pelanggan pada periode tersebut?

## Import Semua Packages/Library yang Digunakan

In [1]:
import pandas as pd

## Data Wrangling

### Gathering Data

In [2]:
# Mengatur data directory
data_dir = 'data/E-Commerce-Public-Dataset'

In [3]:
# Membaca data csv ke dalam dataframe
sellers_df = pd.read_csv(f'{data_dir}/sellers_dataset.csv')
geolocation_df = pd.read_csv(f'{data_dir}/geolocation_dataset.csv')
customers_df = pd.read_csv(f'{data_dir}/customers_dataset.csv')
orders_df = pd.read_csv(f'{data_dir}/orders_dataset.csv')
order_reviews_df = pd.read_csv(f'{data_dir}/order_reviews_dataset.csv')
order_payments_df = pd.read_csv(f'{data_dir}/order_payments_dataset.csv')
order_items_df = pd.read_csv(f'{data_dir}/order_items_dataset.csv')
products_df = pd.read_csv(f'{data_dir}/products_dataset.csv')
product_translations_df = pd.read_csv(f'{data_dir}/product_category_name_translation.csv')

In [4]:
# Melihat 5 data teratas dari dataframe sellers
sellers_df.head()

seller_id  seller_zip_code_prefix  \
0  3442f8959a84dea7ee197c632cb2df15                   13023   
1  d1b65fc7debc3361ea86b5f14c68d2e2                   13844   
2  ce3ad9de960102d0677a81f5d0bb7b2d                   20031   
3  c0f3eea2e14555b6faeea3dd58c1b1c3                    4195   
4  51a04a8a6bdcb23deccc82b0b80742cf                   12914   

         seller_city seller_state  
0           campinas           SP  
1         mogi guacu           SP  
2     rio de janeiro           RJ  
3          sao paulo           SP  
4  braganca paulista           SP

In [5]:
# Melihat 5 data teratas dari dataframe customers
customers_df.head()

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP

In [6]:
# Melihat 5 data teratas dari dataframe orders
orders_df.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  
0           2017-10-18 00:00:00  
1           2018-08-13 00:00:00  
2           2018-09-04 00:00:00  
3           2017-12-15 00:00:00  
4           2018-02-26 00:00:00

In [7]:
# Mengecek tipe data dari setiap kolom di dataframe orders_df
orders_df.dtypes

order_id                         object
customer_id                      object
order_status                     object
order_purchase_timestamp         object
order_approved_at                object
order_delivered_carrier_date     object
order_delivered_customer_date    object
order_estimated_delivery_date    object
dtype: object

In [8]:
# Mengubah kolom yang menunjukkan waktu penanggalan menjadi tipe data datetime
orders_df.order_purchase_timestamp = pd.to_datetime(orders_df.order_purchase_timestamp)
orders_df.order_approved_at = pd.to_datetime(orders_df.order_approved_at)
orders_df.order_delivered_carrier_date = pd.to_datetime(orders_df.order_delivered_carrier_date)
orders_df.order_delivered_customer_date = pd.to_datetime(orders_df.order_delivered_customer_date)
orders_df.order_estimated_delivery_date = pd.to_datetime(orders_df.order_estimated_delivery_date)

In [9]:
# Mengurutkan data berdasarkan waktu pembelian ('order_purchase_timestamp')
orders_df = orders_df.sort_values(by='order_purchase_timestamp')
orders_df.head()

order_id                       customer_id  \
4541   2e7a8482f6fb09756ca50c10d7bfc047  08c5351a6aca1c1589a38f244edeee9d   
4396   e5fa5a7210941f7d56d0208e4e071d35  683c54fc24d40ee9f8a6fc179fd9856c   
10071  809a282bbd5dbcabb6f2f724fca862ec  622e13439d6b5a0b486c435618b2679e   
30710  bfbd0f9bdef84302105ad712db648a6c  86dc2ffce2dfff336de2f386a786e574   
83078  71303d7e93b399f5bcd537d124c0bcfa  b106b360fe2ef8849fbbd056f777b4d5   

      order_status order_purchase_timestamp   order_approved_at  \
4541       shipped      2016-09-04 21:15:19 2016-10-07 13:18:03   
4396      canceled      2016-09-05 00:15:34 2016-10-07 13:17:15   
10071     canceled      2016-09-13 15:24:19 2016-10-07 13:16:46   
30710    delivered      2016-09-15 12:16:38 2016-09-15 12:16:38   
83078     canceled      2016-10-02 22:07:52 2016-10-06 15:50:56   

      order_delivered_carrier_date order_delivered_customer_date  \
4541           2016-10-18 13:14:51                           NaT   
4396                           NaT                           NaT   
10071                          NaT                           NaT   
30710          2016-11-07 17:11:53           2016-11-09 07:47:38   
83078                          NaT                           NaT   

      order_estimated_delivery_date  
4541                     2016-10-20  
4396                     2016-10-28  
10071                    2016-09-30  
30710                    2016-10-04  
83078                    2016-10-25

In [10]:
# Melihat 5 data teratas dari dataframe order_items
order_items_df.head()

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

   shipping_limit_date   price  freight_value  
0  2017-09-19 09:45:35   58.90          13.29  
1  2017-05-03 11:05:13  239.90          19.93  
2  2018-01-18 14:48:30  199.00          17.87  
3  2018-08-15 10:10:18   12.99          12.79  
4  2017-02-13 13:57:51  199.90          18.14

In [11]:
# Mengecek tipe data dari setiap kolom di dataframe order_items_df
order_items_df.dtypes

order_id                object
order_item_id            int64
product_id              object
seller_id               object
shipping_limit_date     object
price                  float64
freight_value          float64
dtype: object

In [12]:
# Mengubah kolom yang menunjukkan waktu penanggalan menjadi tipe data datetime
order_items_df.shipping_limit_date = pd.to_datetime(order_items_df.shipping_limit_date)

In [13]:
# Melihat 5 data teratas dari dataframe products
products_df.head()

product_id  product_category_name  \
0  1e9e8ef04dbcff4541ed26657ea517e5             perfumaria   
1  3aa071139cb16b67ca9e5dea641aaa2f                  artes   
2  96bd76ec8810374ed1b65e291975717f          esporte_lazer   
3  cef67bcfe19066a932b7673e239eb23d                  bebes   
4  9dc1a7de274444849c219cff195d0b71  utilidades_domesticas   

   product_name_lenght  product_description_lenght  product_photos_qty  \
0                 40.0                       287.0                 1.0   
1                 44.0                       276.0                 1.0   
2                 46.0                       250.0                 1.0   
3                 27.0                       261.0                 1.0   
4                 37.0                       402.0                 4.0   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0             225.0               16.0               10.0              14.0  
1            1000.0               30.0               18.0              20.0  
2             154.0               18.0                9.0              15.0  
3             371.0               26.0                4.0              26.0  
4             625.0               20.0               17.0              13.0

In [14]:
# Melihat 5 data teratas dari dataframe product_translations
product_translations_df.head()

product_category_name product_category_name_english
0            beleza_saude                 health_beauty
1  informatica_acessorios         computers_accessories
2              automotivo                          auto
3         cama_mesa_banho                bed_bath_table
4        moveis_decoracao               furniture_decor

In [15]:
# Melihat 5 data teratas dari dataframe order_payments
order_payments_df.head()

order_id  payment_sequential payment_type  \
0  b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1  a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2  25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3  ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4  42fdf880ba16b47b59251dd489d4441a                   1  credit_card   

   payment_installments  payment_value  
0                     8          99.33  
1                     1          24.39  
2                     1          65.71  
3                     8         107.78  
4                     2         128.45

In [16]:
# Melihat 5 data teratas dari dataframe order_reviews
order_reviews_df.head()

review_id                          order_id  \
0  7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1  80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2  228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3  e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4  f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   

   review_score review_comment_title  \
0             4                  NaN   
1             5                  NaN   
2             5                  NaN   
3             5                  NaN   
4             5                  NaN   

                              review_comment_message review_creation_date  \
0                                                NaN  2018-01-18 00:00:00   
1                                                NaN  2018-03-10 00:00:00   
2                                                NaN  2018-02-17 00:00:00   
3              Recebi bem antes do prazo estipulado.  2017-04-21 00:00:00   
4  Parabéns lojas lannister adorei comprar pela I...  2018-03-01 00:00:00   

  review_answer_timestamp  
0     2018-01-18 21:46:59  
1     2018-03-11 03:05:13  
2     2018-02-18 14:36:24  
3     2017-04-21 22:02:06  
4     2018-03-02 10:26:53

In [17]:
# Mengecek tipe data dari setiap kolom di dataframe order_reviews_df
order_reviews_df.dtypes

review_id                  object
order_id                   object
review_score                int64
review_comment_title       object
review_comment_message     object
review_creation_date       object
review_answer_timestamp    object
dtype: object

In [18]:
# Mengubah kolom yang menunjukkan waktu penanggalan menjadi tipe data datetime
order_reviews_df.review_creation_date = pd.to_datetime(order_reviews_df.review_creation_date)
order_reviews_df.review_answer_timestamp = pd.to_datetime(order_reviews_df.review_answer_timestamp)

Berdasarkan informasi dari dataframe di atas maka dapat ditentukan beberapa dataframe yang akan digunakan untuk masing-masing pertanyaan sebagai berikut: 
- Pertanyaan 1 : `orders_df`, `order_items_df`.
- Pertanyaan 2 dan 3 : `orders_df`, `order_items_df`, `products_df`, `product_translations_df`
- Pertanyaan 4 : `orders_df`, `order_items_df`, `products_df`, `product_translations_df`, `order_reviews_df`
- Pertanyaan 5 : `orders_df`, `order_items_df`, `products_df`, `product_translations_df`, `order_reviews_df`, `order_payments_df`

**Insight:**
- Dataset E-Commerce yang digunakan mencakup data pesanan (`orders_df`, `order_reviews_df`, `order_items_df`, `order_payments_df`) dan data produk (`products_df`, `products_translation_df`).

### Assessing Data

#### `df1`

##### Mengecek Data yang Duplikat

In [ ]:
# Menggabungkan dataframe orders dan order_items
df1 = pd.merge(orders_df, order_items_df, on='order_id', how='outer')
df1 = df1.sort_values(by='order_purchase_timestamp', ascending=False)
df1.head()

order_id                       customer_id  \
7392   10a045cdf6a5650c21e9cfeb60384c16  a4b417188addbc05b26b72d5e44837a1   
78023  b059ee4de278302d550a3035c4cdb740  856336203359aa6a61bf3826f7d84c49   
71834  a2ac6dad85cf8af5b0afb510a240fe8c  4c2ec60c29d10c34bd49cb88aa85cfc4   
43094  616fa7d4871b87832197b2a137a115d2  bf6181a85bbb4115736c0a8db1a53be3   
25326  392ed9afd714e3c74767d0c4d3e3f477  2823ffda607a2316375088e0d00005ec   

      order_status order_purchase_timestamp order_approved_at  \
7392      canceled      2018-10-17 17:30:18               NaT   
78023     canceled      2018-10-16 20:16:02               NaT   
71834     canceled      2018-10-03 18:55:29               NaT   
43094     canceled      2018-10-01 15:30:09               NaT   
25326     canceled      2018-09-29 09:13:03               NaT   

      order_delivered_carrier_date order_delivered_customer_date  \
7392                           NaT                           NaT   
78023                          NaT                           NaT   
71834                          NaT                           NaT   
43094                          NaT                           NaT   
25326                          NaT                           NaT   

      order_estimated_delivery_date  order_item_id product_id seller_id  \
7392                     2018-10-30            NaN        NaN       NaN   
78023                    2018-11-12            NaN        NaN       NaN   
71834                    2018-10-16            NaN        NaN       NaN   
43094                    2018-10-23            NaN        NaN       NaN   
25326                    2018-10-15            NaN        NaN       NaN   

      shipping_limit_date  price  freight_value  
7392                  NaT    NaN            NaN  
78023                 NaT    NaN            NaN  
71834                 NaT    NaN            NaN  
43094                 NaT    NaN            NaN  
25326                 NaT    NaN            NaN

In [20]:
# Mengecek data yang duplikat pada df1
df1.duplicated().sum()

0

###### OPSIONAL

In [ ]:
df1[(df1.duplicated(subset=['order_id'], keep=False))].sort_values(by=['order_id','order_item_id'])

In [ ]:
order_check = df1.groupby('order_id').agg({
    'order_item_id': 'nunique',  
    'product_id': 'nunique'  
}).reset_index()

# Ambil order_id yang memiliki lebih dari 1 order_item_id dan lebih dari 1 product_id
filtered_orders = order_check[(order_check['order_item_id'] > 1) & (order_check['product_id'] > 1)]

In [ ]:
filtered_orders

In [ ]:
df1[df1['order_id'].isin(filtered_orders['order_id'])].sort_values(by=['order_id','order_item_id'])

In [ ]:
df1[df1.duplicated(subset=['order_id','product_id'], keep=False)].sort_values(by=['order_id','order_item_id'])

In [ ]:
df1.product_id.nunique()

In [ ]:
df1_clean = df1.dropna(subset=['order_item_id'])
df1_clean = df1_clean.loc[df1_clean.groupby('order_id')['order_item_id'].idxmax()]

In [ ]:
df1_clean[df1_clean.product_id.duplicated()]

In [ ]:
df1_clean[df1_clean.product_id == '28b4eced95a52d9c437a4caf9d311b95']

##### Mengecek Missing Value

In [21]:
# Melihat jumlah missing value pada masing-masing kolom di df1
df1.isnull().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 161
order_delivered_carrier_date     1968
order_delivered_customer_date    3229
order_estimated_delivery_date       0
order_item_id                     775
product_id                        775
seller_id                         775
shipping_limit_date               775
price                             775
freight_value                     775
dtype: int64

Terdapat beberapa kolom yang memiliki missing value. Di antaranya kolom yang perlu dicek adalah kolom 'order_item_id' yang berjumlah 775 dan kolom lainnya yang serupa. Hal ini penting dilakukan karena kolom 'order_item_id' akan digunakan untuk menentukan perkembangan total penjualan yang terjadi di e-commerce ini.

In [22]:
# Melihat data yang memiliki missing value pada kolom order_item_id
df1[df1.order_item_id.isnull()]

order_id                       customer_id  \
7392   10a045cdf6a5650c21e9cfeb60384c16  a4b417188addbc05b26b72d5e44837a1   
78023  b059ee4de278302d550a3035c4cdb740  856336203359aa6a61bf3826f7d84c49   
71834  a2ac6dad85cf8af5b0afb510a240fe8c  4c2ec60c29d10c34bd49cb88aa85cfc4   
43094  616fa7d4871b87832197b2a137a115d2  bf6181a85bbb4115736c0a8db1a53be3   
25326  392ed9afd714e3c74767d0c4d3e3f477  2823ffda607a2316375088e0d00005ec   
...                                 ...                               ...   
73584  a68ce1686d536ca72bd2dadc4b8671e5  d7bed5fac093a4136216072abaf599d5   
98395  ddaec6fff982b13e7e048b627a11d6da  68f4ad79cc0c2ad06e19088f5c00e9fa   
43408  620b0acb9258b51defbc51804c5298d5  9012f48ac2c416c82819101da4e4b75e   
11946  1aecadf4362edaca7fa033e882076c8d  e81a9f176936e3124dfd90c483bf3289   
56938  809a282bbd5dbcabb6f2f724fca862ec  622e13439d6b5a0b486c435618b2679e   

      order_status order_purchase_timestamp   order_approved_at  \
7392      canceled      2018-10-17 17:30:18                 NaT   
78023     canceled      2018-10-16 20:16:02                 NaT   
71834     canceled      2018-10-03 18:55:29                 NaT   
43094     canceled      2018-10-01 15:30:09                 NaT   
25326     canceled      2018-09-29 09:13:03                 NaT   
...            ...                      ...                 ...   
73584      shipped      2016-10-05 01:47:40 2016-10-07 03:11:22   
98395     canceled      2016-10-04 19:41:32                 NaT   
43408     canceled      2016-10-04 11:44:01 2016-10-05 08:45:09   
11946     canceled      2016-10-04 10:05:45 2016-10-04 10:26:40   
56938     canceled      2016-09-13 15:24:19 2016-10-07 13:16:46   

      order_delivered_carrier_date order_delivered_customer_date  \
7392                           NaT                           NaT   
78023                          NaT                           NaT   
71834                          NaT                           NaT   
43094                          NaT                           NaT   
25326                          NaT                           NaT   
...                            ...                           ...   
73584          2016-11-07 16:37:37                           NaT   
98395                          NaT                           NaT   
43408                          NaT                           NaT   
11946                          NaT                           NaT   
56938                          NaT                           NaT   

      order_estimated_delivery_date  order_item_id product_id seller_id  \
7392                     2018-10-30            NaN        NaN       NaN   
78023                    2018-11-12            NaN        NaN       NaN   
71834                    2018-10-16            NaN        NaN       NaN   
43094                    2018-10-23            NaN        NaN       NaN   
25326                    2018-10-15            NaN        NaN       NaN   
...                             ...            ...        ...       ...   
73584                    2016-12-01            NaN        NaN       NaN   
98395                    2016-11-16            NaN        NaN       NaN   
43408                    2016-12-06            NaN        NaN       NaN   
11946                    2016-11-24            NaN        NaN       NaN   
56938                    2016-09-30            NaN        NaN       NaN   

      shipping_limit_date  price  freight_value  
7392                  NaT    NaN            NaN  
78023                 NaT    NaN            NaN  
71834                 NaT    NaN            NaN  
43094                 NaT    NaN            NaN  
25326                 NaT    NaN            NaN  
...                   ...    ...            ...  
73584                 NaT    NaN            NaN  
98395                 NaT    NaN            NaN  
43408                 NaT    NaN            NaN  
11946                 NaT    NaN            NaN  
56938                 NaT    NaN 

In [23]:
# Melihat jumlah data yang memiliki missing value pada kolom order_item_id berdasarkan status pesanan
df1.order_status[df1.order_item_id.isnull()].value_counts()

order_status
unavailable    603
canceled       164
created          5
invoiced         2
shipped          1
Name: count, dtype: int64

`order_item_id` dan kolom lainnya yang memiliki missing value sebanyak 775 disebabkan karena status pesanan yang termasuk : unavailable, canceled, created, invoiced, dan shipped.

In [24]:
df1.order_status[df1.isnull().any(axis=1)].value_counts()

order_status
shipped        1186
canceled        699
unavailable     610
invoiced        361
processing      357
delivered        24
created           5
approved          3
Name: count, dtype: int64

Selain ke-5 status pesanan yang sudah disebutkan sebelumnya, status pesanan lainnya yang memiliki missing value juga diperiksa untuk memastikan data-datanya valid.

In [25]:
df1[(df1.order_status == 'shipped')].isnull().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                   0
order_delivered_carrier_date        0
order_delivered_customer_date    1186
order_estimated_delivery_date       0
order_item_id                       1
product_id                          1
seller_id                           1
shipping_limit_date                 1
price                               1
freight_value                       1
dtype: int64

In [26]:
df1[(df1.order_status == 'shipped') & (df1.order_item_id.isnull())]

order_id                       customer_id  \
73584  a68ce1686d536ca72bd2dadc4b8671e5  d7bed5fac093a4136216072abaf599d5   

      order_status order_purchase_timestamp   order_approved_at  \
73584      shipped      2016-10-05 01:47:40 2016-10-07 03:11:22   

      order_delivered_carrier_date order_delivered_customer_date  \
73584          2016-11-07 16:37:37                           NaT   

      order_estimated_delivery_date  order_item_id product_id seller_id  \
73584                    2016-12-01            NaN        NaN       NaN   

      shipping_limit_date  price  freight_value  
73584                 NaT    NaN            NaN

Berdasarkan informasi di atas, missing value wajar ditemukan pada status pesanan 'shipped' karena pesanan sedang dalam proses pengiriman dan belum sampai ke pelanggan sehingga kolom 'order_delivered_customer_date' masih kosong. Namun ada satu data yang harus dibersihkan karena tidak memiliki informasi jumlah item yang dipesan / 'order_item_id'.

In [27]:
df1[(df1.order_status == 'canceled')].isnull().sum()

order_id                           0
customer_id                        0
order_status                       0
order_purchase_timestamp           0
order_approved_at                141
order_delivered_carrier_date     630
order_delivered_customer_date    699
order_estimated_delivery_date      0
order_item_id                    164
product_id                       164
seller_id                        164
shipping_limit_date              164
price                            164
freight_value                    164
dtype: int64

In [28]:
df1[(df1.order_status == 'unavailable')].isnull().sum()

order_id                           0
customer_id                        0
order_status                       0
order_purchase_timestamp           0
order_approved_at                  0
order_delivered_carrier_date     610
order_delivered_customer_date    610
order_estimated_delivery_date      0
order_item_id                    603
product_id                       603
seller_id                        603
shipping_limit_date              603
price                            603
freight_value                    603
dtype: int64

Pesanan yang di-canceled wajar jika ditemui missing value pada beberapa kolom seperti waktu pengiriman pesanan, jumlah item yang dipesan, dan sebagainya. Begitupula pesanan yang berstatus 'unavailable' juga wajar memiliki missing value pada tanggal kurir menerima pesanan dan tanggal pelanggan menerima pesanan. Kedua status ini bisa dianggap merupakan transaksi yang gagal sehingga informasi terkait jumlah item yang dibeli dan sebagainya tidak ada atau bisa dianggap tidak ada.

In [29]:
df1[(df1.order_status == 'invoiced')].isnull().sum()

order_id                           0
customer_id                        0
order_status                       0
order_purchase_timestamp           0
order_approved_at                  0
order_delivered_carrier_date     361
order_delivered_customer_date    361
order_estimated_delivery_date      0
order_item_id                      2
product_id                         2
seller_id                          2
shipping_limit_date                2
price                              2
freight_value                      2
dtype: int64

In [30]:
df1[(df1.order_status == 'invoiced') & (df1.order_item_id.isnull()) & (df1.isnull().any(axis=1))]

order_id                       customer_id  \
19708  2ce9683175cdab7d1c95bcbb3e36f478  b2d7ae0415dbbca535b5f7b38056dd1f   
99563  e04f1da1f48bf2bbffcf57b9824f76e1  0d00d77134cae4c58695086ad8d85100   

      order_status order_purchase_timestamp   order_approved_at  \
19708     invoiced      2016-10-05 21:03:33 2016-10-06 07:46:39   
99563     invoiced      2016-10-05 13:22:20 2016-10-06 15:51:38   

      order_delivered_carrier_date order_delivered_customer_date  \
19708                          NaT                           NaT   
99563                          NaT                           NaT   

      order_estimated_delivery_date  order_item_id product_id seller_id  \
19708                    2016-11-25            NaN        NaN       NaN   
99563                    2016-11-29            NaN        NaN       NaN   

      shipping_limit_date  price  freight_value  
19708                 NaT    NaN            NaN  
99563                 NaT    NaN            NaN

Pesanan yang berstatus 'invoiced' juga normal memiliki missing value pada kolom order_delivered_carrier_date dan order_delivered_customer_date hanya saja terdapat 2 data yang harus dibersihkan yaitu data yang hilang pada bagian order_item_id

In [31]:
df1[(df1.order_status == 'processing')].isnull().sum()

order_id                           0
customer_id                        0
order_status                       0
order_purchase_timestamp           0
order_approved_at                  0
order_delivered_carrier_date     357
order_delivered_customer_date    357
order_estimated_delivery_date      0
order_item_id                      0
product_id                         0
seller_id                          0
shipping_limit_date                0
price                              0
freight_value                      0
dtype: int64

Pesanan dengan status 'processing' atau dalam proses juga wajar memiliki missing value pada kolom 'order_delivered_carrier_date' dan
'order_delivered_customer_date' sehingga tidak perlu dibersihkan.

In [32]:
df1[(df1.order_status == 'created')].isnull().sum()

order_id                         0
customer_id                      0
order_status                     0
order_purchase_timestamp         0
order_approved_at                5
order_delivered_carrier_date     5
order_delivered_customer_date    5
order_estimated_delivery_date    0
order_item_id                    5
product_id                       5
seller_id                        5
shipping_limit_date              5
price                            5
freight_value                    5
dtype: int64

Pesanan dengan status 'created' juga wajar memiliki missing value. Namun karena informasi penting dari datanya masih kurang cukup seperti missing value pada order_item_id maka kesemua data pesanan dengan status ini tidak digunakan untuk kasus pertanyaan 1.

In [33]:
df1[(df1.order_status == 'approved')].isnull().sum()

order_id                         0
customer_id                      0
order_status                     0
order_purchase_timestamp         0
order_approved_at                0
order_delivered_carrier_date     3
order_delivered_customer_date    3
order_estimated_delivery_date    0
order_item_id                    0
product_id                       0
seller_id                        0
shipping_limit_date              0
price                            0
freight_value                    0
dtype: int64

Pesanan dengan status 'approved' wajar memiliki missing value pada kolom 'order_delivered_carrier_date' dan 'order_delivered_customer_date' sehingga data-datanya sudah valid dan tidak ada yang perlu dibersihkan.

In [34]:
df1[(df1.order_status == 'delivered')].isnull().sum()

order_id                          0
customer_id                       0
order_status                      0
order_purchase_timestamp          0
order_approved_at                15
order_delivered_carrier_date      2
order_delivered_customer_date     8
order_estimated_delivery_date     0
order_item_id                     0
product_id                        0
seller_id                         0
shipping_limit_date               0
price                             0
freight_value                     0
dtype: int64

In [35]:
df1[(df1.order_status == 'delivered') & (df1.isnull().any(axis=1))]

order_id                       customer_id  \
102349  e69f75a717d64fc5ecdfae42b2e8e086  cfda40ca8dd0a5d486a9635b611b398a   
5884    0d3268bad9b086af767785e3f0fc0133  4f1d63d35fb7c8999853b2699f5c7649   
20544   2ebdfc4f15f23b91474edf87475f108e  29f0540231702fda0cfdee0a310f11aa   
14573   20edc82cf5400ce95e1afacc25798b31  28c37425f1127d887d7337f284080a0f   
108941  f5dd62b788049ad9fc0526e3ad11a097  5e89028e024b381dc84a13a3570decb4   
75837   ab7c89dc1bf4a1ead9d6ec1ec8968a84  dd1b84a7286eb4524d52af4256c0ba24   
19784   2d1e2d5bf4dc7227b3bfebb81328c15f  ec05a6d8558c6455f0cbbd8a420ad34f   
18685   2aa91108853cecb43c84a5dc5b277475  afeb16c7f46396c0ed54acb45ccaaa40   
19982   2d858f451373b04fb5c984a1cc2defaf  e08caf668d499a6d643dafd7c5cc498a   
95283   d69e5d356402adc8cf17e08b5033acfb  68d081753ad4fe22fc4d410a9eb1ca01   
60123   88083e8f64d95b932164187484d90212  f67cd1a215aae2a1074638bbd35a223a   
60122   88083e8f64d95b932164187484d90212  f67cd1a215aae2a1074638bbd35a223a   
19141   2babbb4b15e6d2dfe95e2de765c97bce  74bebaf46603f9340e3b50c6b086f992   
41120   5cf925b116421afa85ee25e99b4c34fb  29c35fc91fc13fb5073c8f30505d860d   
36393   51eb2eebd5d76a24625b31c33dd41449  07a2a7e0f63fd8cb757ed77d4245623c   
99553   e04abd8149ef81b95221e88f6ed9ab6a  2127dc6603ac33544953ef05ec155771   
49692   7013bcfc1c97fe719a7b5e05e61c12db  2941af76d38100e0f8740a374f1a5dc3   
61241   8a9adc69528e1001fc68dd0aaebbb54a  4c1ccc74e00993733742a3c786dc3c1f   
95656   d77031d6a3c8a52f019764e68f211c69  0bf35cac6cc7327065da879e2d90fae8   
20625   2eecb0d85f281280f79fa00f9cec1a95  a3d3c38e58b9d2dfb9207cab690b6310   
26595   3c0b8706b065f9919d0505d3b3343881  d85919cb3c0529589c6fa617f5f43281   
8262    12a95a3c06dbaec84bcfb0e2da5d228a  1e101e0daffaddce8159d25a8e53f2b2   
49665   7002a78c79c519ac54022d4f8a65e6e8  d5de688c321096d15508faae67a27051   
85909   c1d4211b3dae76144deccd6c74144a88  684cb238dc5b5d6366244e0e0776b450   

       order_status order_purchase_timestamp   order_approved_at  \
102349    delivered      2018-07-01 22:05:55 2018-07-01 22:15:14   
5884      delivered      2018-07-01 21:14:02 2018-07-01 21:29:54   
20544     delivered      2018-07-01 17:05:11 2018-07-01 17:15:12   
14573     delivered      2018-06-27 16:09:12 2018-06-27 16:29:30   
108941    delivered      2018-06-20 06:58:43 2018-06-20 07:19:05   
75837     delivered      2018-06-08 12:09:39 2018-06-08 12:36:39   
19784     delivered      2017-11-28 17:44:07 2017-11-28 17:56:40   
18685     delivered      2017-09-29 08:52:58 2017-09-29 09:07:16   
19982     delivered      2017-05-25 23:22:43 2017-05-25 23:30:16   
95283     delivered      2017-02-19 01:28:47                 NaT   
60123     delivered      2017-02-18 22:49:19                 NaT   
60122     delivered      2017-02-18 22:49:19                 NaT   
19141     delivered      2017-02-18 17:15:03                 NaT   
41120     delivered      2017-02-18 16:48:35                 NaT   
36393     delivered      2017-02-18 15:52:27                 NaT   
99553     delivered      2017-02-18 14:40:00                 NaT   
49692     delivered      2017-02-18 13:29:47                 NaT   
61241     delivered      2017-02-18 12:45:31                 NaT   
95656     delivered      2017-02-18 11:04:19                 NaT   
20625     delivered      2017-02-17 17:21:55                 NaT   
26595     delivered      2017-02-17 15:53:27                 NaT   
8262      delivered      2017-02-17 13:05:55                 NaT   
49665     delivered      2017-01-19 22:26:59                 NaT   
85909     delivered      2017-01-19 12:48:08                 NaT   

       order_delivered_carrier_date order_delivered_customer_date  \
102349          2018-07-03 13:57:00                           NaT   
5884            2018-07-03 09:28:00                           NaT   
20544           2018-07-03 13:57:00                           NaT   
14573           2018-07-03 19:26:00                           NaT   
108941          2018-06-25 08:05:00

Pada pesanan 'delivered' terdapat beberapa missing value yang dijumpai pada kolom 'order_approved_at', 'order_delivered_carrier_date', dan 'order_delivered_customer_date'. Hal ini bisa disebabkan karena beberapa hal salah satunya adalah kesalahan/error dari aplikasinya itu sendiri. Demi menjaga kevalidan data yang digunakan, maka data-data ini akan dibersihkan.

#### `df23`

##### Mengecek data duplikat

In [ ]:
# Membuat dataframe d23 dengan menggabungkan df1, product_df, dan product_translations_df
df23 = pd.merge(df1, products_df, on='product_id', how='left')
df23 = pd.merge(df23, product_translations_df, on='product_category_name', how='left')
df23 = df23.sort_values(by='order_purchase_timestamp', ascending=False)
df23.head()

order_id                       customer_id  \
0  10a045cdf6a5650c21e9cfeb60384c16  a4b417188addbc05b26b72d5e44837a1   
1  b059ee4de278302d550a3035c4cdb740  856336203359aa6a61bf3826f7d84c49   
2  a2ac6dad85cf8af5b0afb510a240fe8c  4c2ec60c29d10c34bd49cb88aa85cfc4   
3  616fa7d4871b87832197b2a137a115d2  bf6181a85bbb4115736c0a8db1a53be3   
4  392ed9afd714e3c74767d0c4d3e3f477  2823ffda607a2316375088e0d00005ec   

  order_status order_purchase_timestamp order_approved_at  \
0     canceled      2018-10-17 17:30:18               NaT   
1     canceled      2018-10-16 20:16:02               NaT   
2     canceled      2018-10-03 18:55:29               NaT   
3     canceled      2018-10-01 15:30:09               NaT   
4     canceled      2018-09-29 09:13:03               NaT   

  order_delivered_carrier_date order_delivered_customer_date  \
0                          NaT                           NaT   
1                          NaT                           NaT   
2                          NaT                           NaT   
3                          NaT                           NaT   
4                          NaT                           NaT   

  order_estimated_delivery_date  order_item_id product_id  ... freight_value  \
0                    2018-10-30            NaN        NaN  ...           NaN   
1                    2018-11-12            NaN        NaN  ...           NaN   
2                    2018-10-16            NaN        NaN  ...           NaN   
3                    2018-10-23            NaN        NaN  ...           NaN   
4                    2018-10-15            NaN        NaN  ...           NaN   

  product_category_name  product_name_lenght  product_description_lenght  \
0                   NaN                  NaN                         NaN   
1                   NaN                  NaN                         NaN   
2                   NaN                  NaN                         NaN   
3                   NaN                  NaN                         NaN   
4                   NaN                  NaN                         NaN   

  product_photos_qty  product_weight_g  product_length_cm  product_height_cm  \
0                NaN               NaN                NaN                NaN   
1                NaN               NaN                NaN                NaN   
2                NaN               NaN                NaN                NaN   
3                NaN               NaN                NaN                NaN   
4                NaN               NaN                NaN                NaN   

   product_width_cm  product_category_name_english  
0               NaN                            NaN  
1               NaN                            NaN  
2               NaN                            NaN  
3               NaN                            NaN  
4               NaN                            NaN  

[5 rows x 23 columns]

In [ ]:
# Mengecek data yang duplikat pada df23
df23.duplicated().sum()

0

##### Mengecek Missing Value

In [38]:
# Melihat jumlah missing value pada masing-masing kolom di df23
df23.isnull().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 161
order_delivered_carrier_date     1968
order_delivered_customer_date    3229
order_estimated_delivery_date       0
order_item_id                     775
product_id                        775
seller_id                         775
shipping_limit_date               775
price                             775
freight_value                     775
product_category_name            2378
product_name_lenght              2378
product_description_lenght       2378
product_photos_qty               2378
product_weight_g                  793
product_length_cm                 793
product_height_cm                 793
product_width_cm                  793
product_category_name_english    2402
dtype: int64

In [39]:
# Melihat data yang memiliki missing value pada kolom product_category_name
df23[df23.product_category_name.isnull() & df23.product_id.notnull()]

order_id                       customer_id  \
118     e6f6345c50327507fe35965aa65c8e94  b0a44e6002f9ff27862759d0d1992f4e   
178     e53249d8b27242c64053f94e2af2b2ba  94530252be2e605952ea3e82461cfbad   
627     f2fc3e958c7943badc989a1f5db932b4  dbaf23b654a0bbedc2781c69f0c6f03c   
1293    9c33774fba00c73b651903f988c71bb9  21b75a343cc69ed545016d78d22bde39   
1338    61eedf8710394769412bccced6288fff  d37fd6fd3173334ae9deee9569212ced   
...                                  ...                               ...   
112552  22474d84064cd7eb4a907ba3fd8a6398  85f03d59e05ad20dc077b89342b9f710   
112748  6219ece1f053a3ffa649340e87cd93d2  356e49e06716cab5d634234a371dba70   
112840  c022b7da8ab7a4a04bd561c8f765a14a  c797ffdbae65b56c177d3c1368d03628   
113123  b533dea24ada4dc5c1a9041d25d7c812  3a4ccb2c6fb972d301b8f9edfbf81fbd   
113408  0a0837a5eee9e7a9ce2b1fa831944d27  4f3f778022aefa22b9f9e52d2c47edf3   

       order_status order_purchase_timestamp   order_approved_at  \
118       delivered      2018-08-27 18:50:06 2018-08-27 19:05:19   
178       delivered      2018-08-26 21:18:34 2018-08-26 21:35:11   
627       delivered      2018-08-22 21:25:12 2018-08-22 21:35:16   
1293      delivered      2018-08-20 12:37:12 2018-08-21 12:30:50   
1338      delivered      2018-08-20 09:45:10 2018-08-20 14:52:58   
...             ...                      ...                 ...   
112552    delivered      2017-01-25 09:09:46 2017-01-25 09:22:26   
112748    delivered      2017-01-20 15:08:05 2017-01-20 15:21:48   
112840    delivered      2017-01-17 23:40:42 2017-01-17 23:50:15   
113123    delivered      2016-10-08 21:12:48 2016-10-11 04:56:40   
113408    delivered      2016-10-04 09:06:10 2016-10-06 11:43:20   

       order_delivered_carrier_date order_delivered_customer_date  \
118             2018-08-29 15:02:00           2018-08-30 23:16:46   
178             2018-08-27 14:29:00           2018-08-28 22:13:29   
627             2018-08-24 13:34:00           2018-08-27 15:21:34   
1293            2018-08-22 13:35:00           2018-08-30 00:38:24   
1338            2018-08-21 15:37:00           2018-08-27 17:16:52   
...                             ...                           ...   
112552          2017-01-31 10:18:50           2017-02-02 16:56:10   
112748          2017-01-23 10:17:42           2017-02-02 15:15:52   
112840          2017-01-18 13:59:38           2017-01-25 12:51:51   
113123          2016-10-21 12:37:01           2016-10-28 16:31:34   
113408          2016-10-15 10:33:08           2016-10-22 14:51:18   

       order_estimated_delivery_date  order_item_id  \
118                       2018-09-04            1.0   
178                       2018-09-04            1.0   
627                       2018-08-31            1.0   
1293                      2018-08-30            1.0   
1338                      2018-09-04            1.0   
...                              ...            ...   
112552                    2017-03-02            1.0   
112748                    2017-03-08            1.0   
112840                    2017-03-01            1.0   
113123                    2016-12-02            1.0   
113408                    2016-11-24            1.0   

                              product_id  ... freight_value  \
118     fd85f34d70edf41fbea6a3269d183e00  ...          7.47   
178     b0645512783cc56c9510f08e90a40ede  ...          7.68   
627     6ab9d49c399f239bc88ea28c7568213f  ...          7.39   
1293    f524a09994004b2dabb3d13cf1a041fb  ...         42.73   
1338    5a848e4ab52fd5445cdc07aab1c40e48  ...         15.96   
...                                  ...  ...           ...   
112552  6f6f3b1a96452e2fbad0cd4f7082765b  ...         16.88   
112748  aa2b29bee083c3eb3779f39fc09527b7  ...         20.69   
112840  d36c8237f8a0a889aa9d6ac7816097e3  ...         25.00   
113123  661e3bba6bdb874638367462b49ae590  ...         14.52   
113408  583916a5dae918f5e89baec139141c54  ...         15.51   

       product_category_name  p

Terdapat 1.603 data product_id yang tidak memiliki product_category_name. 

In [40]:
df23.product_category_name[df23.product_category_name.notnull() & df23.product_category_name_english.isnull()].value_counts()

product_category_name
portateis_cozinha_e_preparadores_de_alimentos    15
pc_gamer                                          9
Name: count, dtype: int64

Terdapat 24 data yang memiliki product_category_name tetapi tidak ada di product_category_name_english. Untuk mengatasi masalah ini maka data tersebut dapat diterjemahkan secara manual ke bahasa inggris lalu digabungkan ke kategori produk yang mendekati.
- portateis_cozinha_e_preparandores_de_alimentos : portable_kitchen_and_food_preparers
- pc_gamer : pc_gamer

#### `df4`

##### Mengecek Data Duplikat

In [ ]:
# Membuat dataframe df4 dengan menggabungkan dataframe df23 dengan order_reviews_df
df4 = pd.merge(df23, order_reviews_df, on='order_id', how='left')
df4 = df4.sort_values(by='order_purchase_timestamp', ascending=False)
df4.head()

order_id                       customer_id  \
0  10a045cdf6a5650c21e9cfeb60384c16  a4b417188addbc05b26b72d5e44837a1   
1  b059ee4de278302d550a3035c4cdb740  856336203359aa6a61bf3826f7d84c49   
2  a2ac6dad85cf8af5b0afb510a240fe8c  4c2ec60c29d10c34bd49cb88aa85cfc4   
3  616fa7d4871b87832197b2a137a115d2  bf6181a85bbb4115736c0a8db1a53be3   
4  392ed9afd714e3c74767d0c4d3e3f477  2823ffda607a2316375088e0d00005ec   

  order_status order_purchase_timestamp order_approved_at  \
0     canceled      2018-10-17 17:30:18               NaT   
1     canceled      2018-10-16 20:16:02               NaT   
2     canceled      2018-10-03 18:55:29               NaT   
3     canceled      2018-10-01 15:30:09               NaT   
4     canceled      2018-09-29 09:13:03               NaT   

  order_delivered_carrier_date order_delivered_customer_date  \
0                          NaT                           NaT   
1                          NaT                           NaT   
2                          NaT                           NaT   
3                          NaT                           NaT   
4                          NaT                           NaT   

  order_estimated_delivery_date  order_item_id product_id  ...  \
0                    2018-10-30            NaN        NaN  ...   
1                    2018-11-12            NaN        NaN  ...   
2                    2018-10-16            NaN        NaN  ...   
3                    2018-10-23            NaN        NaN  ...   
4                    2018-10-15            NaN        NaN  ...   

  product_length_cm product_height_cm  product_width_cm  \
0               NaN               NaN               NaN   
1               NaN               NaN               NaN   
2               NaN               NaN               NaN   
3               NaN               NaN               NaN   
4               NaN               NaN               NaN   

   product_category_name_english                         review_id  \
0                            NaN  716eff9ef436f351494c303085fcfcf5   
1                            NaN  48198628b2d43ed394089750acc59c3f   
2                            NaN  4548534449b1f572e357211b90724f1b   
3                            NaN  8b8f495d30fca09fdc8312e96de1bc47   
4                            NaN  f1f8696d256889c0fe9962a5d32746e0   

   review_score  review_comment_title  \
0           1.0                Troca    
1           5.0                Gostei   
2           1.0                   NaN   
3           2.0                   NaN   
4           1.0                   NaN   

                              review_comment_message  review_creation_date  \
0  Por gentileza, trocar meu produto pois comprei...            2018-08-23   
1  Os produtos são de boa qualidade e o preço que...            2018-06-27   
2  quero devolver o produto comprei um veio outro...            2018-08-18   
3                                                NaN            2018-08-23   
4  o produto foi entregue errado eu estou esperad...            2018-08-26   

   review_answer_timestamp  
0      2018-08-24 02:56:26  
1      2018-06-30 01:48:53  
2      2018-08-19 00:39:54  
3      2018-08-23 23:14:33  
4      2018-08-28 01:26:38  

[5 rows x 29 columns]

In [ ]:
# Menghitung jumlah data yang duplikat setiap kolom di df4
df4.duplicated().sum()

0

##### Mengecek Missing Value

In [43]:
# Menghitung jumlah missing value pada masing-masing kolom di df4
df4.isnull().sum()

order_id                              0
customer_id                           0
order_status                          0
order_purchase_timestamp              0
order_approved_at                   162
order_delivered_carrier_date       1980
order_delivered_customer_date      3253
order_estimated_delivery_date         0
order_item_id                       778
product_id                          778
seller_id                           778
shipping_limit_date                 778
price                               778
freight_value                       778
product_category_name              2390
product_name_lenght                2390
product_description_lenght         2390
product_photos_qty                 2390
product_weight_g                    796
product_length_cm                   796
product_height_cm                   796
product_width_cm                    796
product_category_name_english      2414
review_id                           961
review_score                        961


In [48]:
# Persentase missing value pada kolom review_comment_title
print(f'{df4.review_comment_title.isnull().sum() / len(df4) * 100:.2f}%')

88.15%


In [49]:
# Persentase missing value pada kolom review_comment_message
print(f'{df4.review_comment_message.isnull().sum() / len(df4) * 100:.2f}%')

57.78%


In [ ]:
df4[df4.review_score.isnull()]

order_id                       customer_id  \
15      afb61112fb99b07fe17e591c68c0c84c  36b182c39933d55e307096ccc3fd2970   
125     e91a3bc14b8d125d82f0334371c39926  befe184e4eb181470e6063a1a368a093   
1344    37e5725572d3cbd7aad103b730f9232a  3dd789d5301761182d28eb73ce1558ed   
1596    aa6e3cc7cf13ef1dc599d1243daee6f6  af18612bfb6a3f68257c693f72848f28   
1825    4ac05edf248d6206469e17c32e52c9ac  741f00aca12b97beb2db34f629655c30   
...                                  ...                               ...   
113849  0efd0bc268d34da3f01f4ff25a6d4335  95446917717bb58d553d107d0f1668f6   
113864  02190241f7190a1f3c7e0df95a749c6a  55c9044401ffab0c717f24405e2e042f   
113865  02190241f7190a1f3c7e0df95a749c6a  55c9044401ffab0c717f24405e2e042f   
113952  5da2004d85f2828349d887c989da2adc  95f592086150f0c95d3082aea047b816   
113970  cda873529ca7ab71f677d5ec11a40304  76c74aaff2f3f7355f46d9818ad092b8   

       order_status order_purchase_timestamp   order_approved_at  \
15         canceled      2018-09-06 10:48:12                 NaT   
125       delivered      2018-08-27 17:01:43 2018-08-27 17:10:27   
1344      delivered      2018-08-20 09:18:49 2018-08-20 14:53:41   
1596      delivered      2018-08-18 23:54:05 2018-08-21 04:15:19   
1825      delivered      2018-08-17 22:35:51 2018-08-17 22:49:33   
...             ...                      ...                 ...   
113849      shipped      2016-10-07 15:53:31 2016-10-07 16:12:22   
113864    delivered      2016-10-07 10:52:22 2016-10-08 04:44:22   
113865    delivered      2016-10-07 10:52:22 2016-10-08 04:44:22   
113952    delivered      2016-10-05 20:20:59 2016-10-07 02:48:26   
113970      shipped      2016-10-05 16:57:30 2016-10-06 15:52:49   

       order_delivered_carrier_date order_delivered_customer_date  \
15                              NaT                           NaT   
125             2018-08-28 14:56:00           2018-08-29 20:04:52   
1344            2018-08-28 16:13:00           2018-08-29 22:58:53   
1596            2018-08-21 14:21:00           2018-08-27 18:40:41   
1825            2018-08-22 16:33:00           2018-08-25 12:07:33   
...                             ...                           ...   
113849          2016-10-11 16:12:23                           NaT   
113864          2016-10-21 12:47:02           2016-10-27 09:37:52   
113865          2016-10-21 12:47:02           2016-10-27 09:37:52   
113952          2016-11-21 10:23:41           2016-11-26 08:16:49   
113970          2016-11-14 11:14:39                           NaT   

       order_estimated_delivery_date  order_item_id  \
15                        2018-09-19            NaN   
125                       2018-08-31            1.0   
1344                      2018-08-27            1.0   
1596                      2018-09-20            1.0   
1825                      2018-09-12            1.0   
...                              ...            ...   
113849                    2016-12-13            1.0   
113864                    2016-11-29            2.0   
113865                    2016-11-29            1.0   
113952                    2016-11-29            1.0   
113970                    2016-12-01            1.0   

                              product_id  ... product_length_cm  \
15                                   NaN  ...               NaN   
125     2715967684d6991f7501693bfe08c08b  ...              28.0   
1344    203086d8efe5991e11ed8f673bd18d64  ...              25.0   
1596    66740f6bc8c6a2393283871b22d5d32f  ...              40.0   
1825    2ee7612eb9c995c7f39c6f8ae967e1a9  ...              16.0   
...                                  ...  ...               ...   
113849  6dab56beb5263f0d554cae5c55809208  ...              16.0   
113864  e00d2c49504aeaf7acdc9ff7cf8a6107  ...              30.0   
113865  e00d2c49504aeaf7acdc9ff7cf8a6107  ...              30.0   
113952  72bf8fd8b4596a7fa2c2cfab031204ed  ...              19.0   
113970  79938780eb7ffb7289f59a9a283bc06b  ...      

Pada pertanyaan 4, fokus utamanya adalah nilai review_score yang tidak boleh mengandung missing_value. Berdasarkan informasi di atas terdapat 961 data missing value pada df4 berdasarkan review_score. Selain itu jika kita melihat kolom pada 'review_comment_title' dan 'review_comment_message' jumlah missing valuenya terlalu banyak sehingga kolom ini akan dihapus.

#### `df5`

##### Mengecek Data Duplikat

In [ ]:
# Membuat dataframe df5 dengan menggabungkan dataframe df4 dengan order_payments_df
df5 = pd.merge(df4, order_payments_df, on='order_id', how='left')
df5.head()

order_id                       customer_id  \
0  10a045cdf6a5650c21e9cfeb60384c16  a4b417188addbc05b26b72d5e44837a1   
1  b059ee4de278302d550a3035c4cdb740  856336203359aa6a61bf3826f7d84c49   
2  a2ac6dad85cf8af5b0afb510a240fe8c  4c2ec60c29d10c34bd49cb88aa85cfc4   
3  616fa7d4871b87832197b2a137a115d2  bf6181a85bbb4115736c0a8db1a53be3   
4  392ed9afd714e3c74767d0c4d3e3f477  2823ffda607a2316375088e0d00005ec   

  order_status order_purchase_timestamp order_approved_at  \
0     canceled      2018-10-17 17:30:18               NaT   
1     canceled      2018-10-16 20:16:02               NaT   
2     canceled      2018-10-03 18:55:29               NaT   
3     canceled      2018-10-01 15:30:09               NaT   
4     canceled      2018-09-29 09:13:03               NaT   

  order_delivered_carrier_date order_delivered_customer_date  \
0                          NaT                           NaT   
1                          NaT                           NaT   
2                          NaT                           NaT   
3                          NaT                           NaT   
4                          NaT                           NaT   

  order_estimated_delivery_date  order_item_id product_id  ...  \
0                    2018-10-30            NaN        NaN  ...   
1                    2018-11-12            NaN        NaN  ...   
2                    2018-10-16            NaN        NaN  ...   
3                    2018-10-23            NaN        NaN  ...   
4                    2018-10-15            NaN        NaN  ...   

                          review_id review_score  review_comment_title  \
0  716eff9ef436f351494c303085fcfcf5          1.0                Troca    
1  48198628b2d43ed394089750acc59c3f          5.0                Gostei   
2  4548534449b1f572e357211b90724f1b          1.0                   NaN   
3  8b8f495d30fca09fdc8312e96de1bc47          2.0                   NaN   
4  f1f8696d256889c0fe9962a5d32746e0          1.0                   NaN   

                              review_comment_message review_creation_date  \
0  Por gentileza, trocar meu produto pois comprei...           2018-08-23   
1  Os produtos são de boa qualidade e o preço que...           2018-06-27   
2  quero devolver o produto comprei um veio outro...           2018-08-18   
3                                                NaN           2018-08-23   
4  o produto foi entregue errado eu estou esperad...           2018-08-26   

   review_answer_timestamp  payment_sequential  payment_type  \
0      2018-08-24 02:56:26                 1.0       voucher   
1      2018-06-30 01:48:53                 1.0       voucher   
2      2018-08-19 00:39:54                 1.0       voucher   
3      2018-08-23 23:14:33                 1.0       voucher   
4      2018-08-28 01:26:38                 1.0       voucher   

   payment_installments  payment_value  
0                   1.0          89.71  
1                   1.0         222.03  
2                   1.0         197.55  
3                   1.0          80.38  
4                   1.0         137.03  

[5 rows x 33 columns]

In [52]:
# Menghitung jumlah data yang duplikat pada setiap kolom di df5
df5.duplicated().sum()

0

##### Mengecek Missing Value

In [53]:
# Menghitung jumlah missing value pada masing-masing kolom di df5
df5.isnull().sum()

order_id                              0
customer_id                           0
order_status                          0
order_purchase_timestamp              0
order_approved_at                   177
order_delivered_carrier_date       2086
order_delivered_customer_date      3421
order_estimated_delivery_date         0
order_item_id                       833
product_id                          833
seller_id                           833
shipping_limit_date                 833
price                               833
freight_value                       833
product_category_name              2542
product_name_lenght                2542
product_description_lenght         2542
product_photos_qty                 2542
product_weight_g                    853
product_length_cm                   853
product_height_cm                   853
product_width_cm                    853
product_category_name_english      2567
review_id                           997
review_score                        997


Sama halnya dengan pertanyaan ke-4, pertanyaan ke-5 fokus utamanya adalah nilai review_score yang tidak boleh mengandung missing_value dan faktor-faktor yang menyebabkan hal tersebut bisa terjadi.

In [54]:
df5[df5.review_score.isnull()]

order_id                       customer_id  \
15      afb61112fb99b07fe17e591c68c0c84c  36b182c39933d55e307096ccc3fd2970   
128     e91a3bc14b8d125d82f0334371c39926  befe184e4eb181470e6063a1a368a093   
1369    37e5725572d3cbd7aad103b730f9232a  3dd789d5301761182d28eb73ce1558ed   
1630    aa6e3cc7cf13ef1dc599d1243daee6f6  af18612bfb6a3f68257c693f72848f28   
1866    4ac05edf248d6206469e17c32e52c9ac  741f00aca12b97beb2db34f629655c30   
...                                  ...                               ...   
118889  0efd0bc268d34da3f01f4ff25a6d4335  95446917717bb58d553d107d0f1668f6   
118904  02190241f7190a1f3c7e0df95a749c6a  55c9044401ffab0c717f24405e2e042f   
118905  02190241f7190a1f3c7e0df95a749c6a  55c9044401ffab0c717f24405e2e042f   
118992  5da2004d85f2828349d887c989da2adc  95f592086150f0c95d3082aea047b816   
119010  cda873529ca7ab71f677d5ec11a40304  76c74aaff2f3f7355f46d9818ad092b8   

       order_status order_purchase_timestamp   order_approved_at  \
15         canceled      2018-09-06 10:48:12                 NaT   
128       delivered      2018-08-27 17:01:43 2018-08-27 17:10:27   
1369      delivered      2018-08-20 09:18:49 2018-08-20 14:53:41   
1630      delivered      2018-08-18 23:54:05 2018-08-21 04:15:19   
1866      delivered      2018-08-17 22:35:51 2018-08-17 22:49:33   
...             ...                      ...                 ...   
118889      shipped      2016-10-07 15:53:31 2016-10-07 16:12:22   
118904    delivered      2016-10-07 10:52:22 2016-10-08 04:44:22   
118905    delivered      2016-10-07 10:52:22 2016-10-08 04:44:22   
118992    delivered      2016-10-05 20:20:59 2016-10-07 02:48:26   
119010      shipped      2016-10-05 16:57:30 2016-10-06 15:52:49   

       order_delivered_carrier_date order_delivered_customer_date  \
15                              NaT                           NaT   
128             2018-08-28 14:56:00           2018-08-29 20:04:52   
1369            2018-08-28 16:13:00           2018-08-29 22:58:53   
1630            2018-08-21 14:21:00           2018-08-27 18:40:41   
1866            2018-08-22 16:33:00           2018-08-25 12:07:33   
...                             ...                           ...   
118889          2016-10-11 16:12:23                           NaT   
118904          2016-10-21 12:47:02           2016-10-27 09:37:52   
118905          2016-10-21 12:47:02           2016-10-27 09:37:52   
118992          2016-11-21 10:23:41           2016-11-26 08:16:49   
119010          2016-11-14 11:14:39                           NaT   

       order_estimated_delivery_date  order_item_id  \
15                        2018-09-19            NaN   
128                       2018-08-31            1.0   
1369                      2018-08-27            1.0   
1630                      2018-09-20            1.0   
1866                      2018-09-12            1.0   
...                              ...            ...   
118889                    2016-12-13            1.0   
118904                    2016-11-29            2.0   
118905                    2016-11-29            1.0   
118992                    2016-11-29            1.0   
119010                    2016-12-01            1.0   

                              product_id  ... review_id review_score  \
15                                   NaN  ...       NaN          NaN   
128     2715967684d6991f7501693bfe08c08b  ...       NaN          NaN   
1369    203086d8efe5991e11ed8f673bd18d64  ...       NaN          NaN   
1630    66740f6bc8c6a2393283871b22d5d32f  ...       NaN          NaN   
1866    2ee7612eb9c995c7f39c6f8ae967e1a9  ...       NaN          NaN   
...                                  ...  ...       ...          ...   
118889  6dab56beb5263f0d554cae5c55809208  ...       NaN          NaN   
118904  e00d2c49504aeaf7acdc9ff7cf8a6107  ...       NaN          NaN   
118905  e00d2c49504aeaf7acdc9ff7cf8a6107  ...       NaN          NaN   
118992  72bf8fd8b4596a7fa2c2cfab031204ed  ...       NaN          NaN

Terdapat 997 data missing value yang ditemukan. Jumlah data missing value ini lebih banyak dibanding `df4` dan bisa disebabkan karena jenis payment_method yang digunakan pelanggan beragam atau jumlah order_item_id yang lebih dari satu juga berpengaruh.

#### SEBELUM MERGE

In [ ]:
# Mengecek dataframe yang memiliki nilai duplikat
print('Dataframe\t\t Jumlah Duplikat')
print('orders_df\t\t:', orders_df.duplicated().sum())
print('order_items_df\t\t:', order_items_df.duplicated().sum())
print('products_df\t\t:', products_df.duplicated().sum())
print('products_translation_df\t:', products_translation_df.duplicated().sum())
print('order_reviews_df\t:', order_reviews_df.duplicated().sum())
print('order_payments_df\t:', order_payments_df.duplicated().sum())
print('customers_df\t\t:', customers_df.duplicated().sum())
print('sellers_df\t\t:', sellers_df.duplicated().sum())

In [ ]:
# Melihat data yang duplikat
geolocation_df[geolocation_df.duplicated()]

In [ ]:
# Contoh data yang duplikat pada geolocation_df
geolocation_df[(geolocation_df.geolocation_zip_code_prefix == 1050) & (geolocation_df.geolocation_lat == -23.54985418067981) & (geolocation_df.geolocation_lng == -46.643139396906825)]

In [ ]:
# Mengecek Missing Value pada Dataframe
print('Dataframe\t\t Jumlah Missing Value')
print('orders_df\t\t:', orders_df.isnull().sum().sum())
print('order_items_df\t\t:', order_items_df.isnull().sum().sum())
print('products_df\t\t:', products_df.isnull().sum().sum())
print('products_translation_df\t:', products_translation_df.isnull().sum().sum())
print('order_reviews_df\t:', order_reviews_df.isnull().sum().sum())
print('order_payments_df\t:', order_payments_df.isnull().sum().sum())
print('customers_df\t\t:', customers_df.isnull().sum().sum())
print('sellers_df\t\t:', sellers_df.isnull().sum().sum())
print('geolocation_df\t\t:', geolocation_df.isnull().sum().sum())

**Insight:**
- `geolocation_df` memiliki 261.831 data duplikat
- `orders_df` memiliki 4.908 missing value (`order_approved_at`: 160, `order_delivered_carrier_date`: 1.783, `order_delivered_customer_date`: 2.965)
- `products_df` memiliki 2.448 missing value (`product_category_name`: 610, `product_name_lenght`: 610, `product_description_lenght`: 610, `product_photos_qty`: 610, `product_weight_g`: 2, `product_length_cm`: 2, `product_height_cm`: 2, `product_width_cm`: 2)
- `order_reviews_df` memiliki 145.903 missing value (`review_comment_title`: 87.656, `review_comment_message`: 58.247)

### Cleaning Data

#### `df1`

Berdasarkan informasi yang sudah didapat saat assessing data `df1` maka yang perlu dibersihkan adalah :
- Data-data yang kolom 'order_item_id' nya kosong/missing value
- Menghapus data-data dengan status pesanan 'canceled' dan 'unvailable'
- Menghapus data-data dengan status pesanan 'delivered' yang datanya terdapat missing value

In [101]:
# Menghapus baris data yang mengandung missing value pada kolom order_item_id
df1.dropna(subset=['order_item_id'], inplace=True)

In [102]:
# Menghapus baris data dengan status pesan 'canceled' dan 'unavailable'
df1.drop(df1[df1.order_status == 'canceled'].index, inplace=True)
df1.drop(df1[df1.order_status == 'unavailable'].index, inplace=True)

In [103]:
# Menghapus baris data dengan status pesan 'delivered' yang memiliki missing value
df1.drop(df1[(df1.order_status == 'delivered') & (df1.isnull().any(axis=1))].index, inplace=True)

In [104]:
df1.order_status.value_counts()

order_status
delivered     110173
shipped         1185
invoiced         359
processing       357
approved           3
Name: count, dtype: int64

In [105]:
df1[df1.order_status == 'delivered'].isnull().sum()

order_id                         0
customer_id                      0
order_status                     0
order_purchase_timestamp         0
order_approved_at                0
order_delivered_carrier_date     0
order_delivered_customer_date    0
order_estimated_delivery_date    0
order_item_id                    0
product_id                       0
seller_id                        0
shipping_limit_date              0
price                            0
freight_value                    0
dtype: int64

In [106]:
df1[df1.order_status == 'shipped'].isnull().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                   0
order_delivered_carrier_date        0
order_delivered_customer_date    1185
order_estimated_delivery_date       0
order_item_id                       0
product_id                          0
seller_id                           0
shipping_limit_date                 0
price                               0
freight_value                       0
dtype: int64

In [107]:
df1[df1.order_status == 'canceled'].isnull().sum()

order_id                         0
customer_id                      0
order_status                     0
order_purchase_timestamp         0
order_approved_at                0
order_delivered_carrier_date     0
order_delivered_customer_date    0
order_estimated_delivery_date    0
order_item_id                    0
product_id                       0
seller_id                        0
shipping_limit_date              0
price                            0
freight_value                    0
dtype: int64

In [108]:
df1[df1.order_status == 'invoiced'].isnull().sum()

order_id                           0
customer_id                        0
order_status                       0
order_purchase_timestamp           0
order_approved_at                  0
order_delivered_carrier_date     359
order_delivered_customer_date    359
order_estimated_delivery_date      0
order_item_id                      0
product_id                         0
seller_id                          0
shipping_limit_date                0
price                              0
freight_value                      0
dtype: int64

In [109]:
df1[df1.order_status == 'processing'].isnull().sum()

order_id                           0
customer_id                        0
order_status                       0
order_purchase_timestamp           0
order_approved_at                  0
order_delivered_carrier_date     357
order_delivered_customer_date    357
order_estimated_delivery_date      0
order_item_id                      0
product_id                         0
seller_id                          0
shipping_limit_date                0
price                              0
freight_value                      0
dtype: int64

In [110]:
df1[df1.order_status == 'approved'].isnull().sum()

order_id                         0
customer_id                      0
order_status                     0
order_purchase_timestamp         0
order_approved_at                0
order_delivered_carrier_date     3
order_delivered_customer_date    3
order_estimated_delivery_date    0
order_item_id                    0
product_id                       0
seller_id                        0
shipping_limit_date              0
price                            0
freight_value                    0
dtype: int64

In [111]:
# Melihat jumlah missing value pada masing-masing kolom di df1
df1.isnull().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                   0
order_delivered_carrier_date      719
order_delivered_customer_date    1904
order_estimated_delivery_date       0
order_item_id                       0
product_id                          0
seller_id                           0
shipping_limit_date                 0
price                               0
freight_value                       0
dtype: int64

**Insight:**
- Dataframe `df1` sudah memiliki data yang valid berdasarkan status pesanannya.
- Data yang hilang/missing value pada kolom 'order_item_id' sudah dibersihkan.

#### `df23`

Berdasarkan informasi yang didapat dari proses assessing data beberapa hal yang perlu dibersihkan/dimanipulasi datanya sebagai berikut:
- Data-data yang tidak memiliki product_category_name
- Data-data yang tidak memiliki nilai di product_category_name_english

In [112]:
# Memperbarui dataframe d23 dengan data df1 yang sudah dibersihkan
df23 = pd.merge(df1, products_df, on='product_id', how='left')
df23 = pd.merge(df23, product_translations_df, on='product_category_name', how='left')
df23 = df23.sort_values(by='order_purchase_timestamp', ascending=False)
df23.head()

order_id                       customer_id  \
0  54282e97f61c23b78330c15b154c867d  4b7decb9b58e2569548b8b4c8e20e8d7   
1  35a972d7f8436f405b56e36add1a7140  898b7fee99c4e42170ab69ba59be0a8b   
2  03ef5dedbe7492bdae72eec50764c43f  496630b6740bcca28fce9ba50d8a26ef   
3  168626408cb32af0ffaf76711caae1dc  6e353700bc7bcdf6ebc15d6de16d7002   
4  0b223d92c27432930dfe407c6aea3041  e60df9449653a95af4549bbfcb18a6eb   

  order_status order_purchase_timestamp   order_approved_at  \
0      shipped      2018-09-03 09:06:57 2018-09-03 17:40:06   
1    delivered      2018-08-29 15:00:37 2018-08-29 15:10:26   
2    delivered      2018-08-29 14:52:00 2018-08-29 15:05:22   
3    delivered      2018-08-29 14:18:28 2018-08-29 14:30:23   
4    delivered      2018-08-29 14:18:23 2018-08-29 14:31:07   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2018-09-04 15:25:00                           NaT   
1          2018-08-29 16:57:00           2018-08-30 16:23:36   
2          2018-08-29 20:01:00           2018-08-30 16:36:59   
3          2018-08-29 18:51:00           2018-08-30 16:52:31   
4          2018-08-29 15:29:00           2018-08-30 16:24:55   

  order_estimated_delivery_date  order_item_id  \
0                    2018-09-06            1.0   
1                    2018-09-05            1.0   
2                    2018-09-03            1.0   
3                    2018-09-11            1.0   
4                    2018-09-04            1.0   

                         product_id  ... freight_value  \
0  b98992ea80b467987a7fbb88e7f2076a  ...         21.46   
1  d04857e7b4b708ee8b8b9921163edba3  ...          8.76   
2  c7f27c5bef2338541c772b5776403e6a  ...          8.33   
3  bdcf6a834e8faa30dac3886c7a58e92e  ...         15.39   
4  2b4472df15512a2825ae86fd9ae79335  ...         46.48   

                            product_category_name  product_name_lenght  \
0  moveis_cozinha_area_de_servico_jantar_e_jardim                 58.0   
1                          informatica_acessorios                 33.0   
2                               artigos_de_festas                 54.0   
3                                    beleza_saude                 26.0   
4  moveis_cozinha_area_de_servico_jantar_e_jardim                 44.0   

   product_description_lenght product_photos_qty  product_weight_g  \
0                       576.0                1.0           10400.0   
1                      1127.0                2.0             450.0   
2                       429.0                1.0             450.0   
3                       394.0                1.0            1614.0   
4                       112.0                1.0           13550.0   

   product_length_cm  product_height_cm  product_width_cm  \
0               16.0               68.0              32.0   
1               16.0                4.0              20.0   
2               28.0                9.0              19.0   
3               31.0               16.0              28.0   
4               48.0               49.0              49.0   

             product_category_name_english  
0  kitchen_dining_laundry_garden_furniture  
1                    computers_accessories  
2                           party_supplies  
3                            health_beauty  
4  kitchen_dining_laundry_garden_furniture  

[5 rows x 23 columns]

In [113]:
# Menghapus baris data yang memiliki missing value pada kolom product_category_name
df23.drop(df23[df23.product_category_name.isnull()].index, inplace=True)

In [114]:
# Melihat data yang memiliki missing value pada kolom product_category_name_english 
df23.product_category_name[df23.product_category_name_english.isnull()].value_counts()

product_category_name
portateis_cozinha_e_preparadores_de_alimentos    14
pc_gamer                                          8
Name: count, dtype: int64

In [115]:
df23.loc[(df23.product_category_name == 'portateis_cozinha_e_preparadores_de_alimentos'), 'product_category_name_english'] = 'portable_kitchen_and_food_preparers'

In [116]:
df23[df23.product_category_name == 'portateis_cozinha_e_preparadores_de_alimentos']

order_id                       customer_id  \
115    a73d469a8747dea1740e9545d1051002  fa5e5ede73e8f02fe888b6f3bd44a714   
1868   0745fd0c5e5bd55f752798a152b1d04b  c24fc4f24d2c6bd64d612774f52d6a97   
3955   8dbdb81783a3c754dc9441826510a36a  2d0c34b4b379c290f079d39df12304e2   
5566   2d3bc1f6ed458a137c51adc3cab7a488  e620a9a4cdf12ab1e817c17496e9c5d7   
8389   67f38c0b9d8a42541af92de7b6a19ea5  46ae9e1a0e2b936715ed2634a524f8ab   
8424   d94600e5b855ffe8db3b9bc06e21b1bd  9f35486803ac72d115dedc28ef6156e1   
8425   d94600e5b855ffe8db3b9bc06e21b1bd  9f35486803ac72d115dedc28ef6156e1   
8448   53fa17c349c4b3dcbbadd8aad2eb559b  de4347cfed448a35884a19949bc53087   
9778   b35a771164a9defa3b4767f1b66918b1  db72b1bc11dfa3a7d4dc04332a77b651   
9940   c67c6e7178cff9848b2d6556d1670fd8  3237a9ea092013b7b44c0c8d5174006d   
10297  542dd8c7a80f7006b56c9cbb95e6433b  416d7934b96e1c5d46b332d32ee0e356   
10505  b1829b9fc61d2f28d8668cd5bde29e3f  bd371cfbb78aa6bd8d2ad4dc52565d8f   
26600  a653bb2e2418fd13d746e3639879aa83  404e9aa29dd2f4f68be287245f0e1e50   
26838  7faf8b9448b66f4a3ac98a312b3e7609  c0cc59cb5d494116de580acf7e6c46ac   

      order_status order_purchase_timestamp   order_approved_at  \
115      delivered      2018-08-27 12:41:49 2018-08-27 13:15:50   
1868     delivered      2018-08-17 15:45:42 2018-08-21 04:30:19   
3955     delivered      2018-08-10 14:54:25 2018-08-10 15:10:27   
5566     delivered      2018-08-06 09:21:25 2018-08-06 09:35:13   
8389     delivered      2018-07-27 16:59:35 2018-07-27 17:15:17   
8424     delivered      2018-07-27 14:33:28 2018-07-28 03:35:16   
8425     delivered      2018-07-27 14:33:28 2018-07-28 03:35:16   
8448     delivered      2018-07-27 13:09:23 2018-07-27 13:43:56   
9778     delivered      2018-07-23 00:49:15 2018-07-25 02:50:18   
9940     delivered      2018-07-22 13:32:50 2018-07-23 11:32:05   
10297    delivered      2018-07-20 17:53:08 2018-07-20 18:05:18   
10505    delivered      2018-07-20 01:38:27 2018-07-20 01:50:13   
26600    delivered      2018-05-09 10:34:53 2018-05-09 10:54:54   
26838    delivered      2018-05-08 17:18:14 2018-05-08 17:35:18   

      order_delivered_carrier_date order_delivered_customer_date  \
115            2018-08-28 15:07:00           2018-08-30 22:24:46   
1868           2018-08-21 14:53:00           2018-08-30 16:08:27   
3955           2018-08-13 14:00:00           2018-08-17 20:51:49   
5566           2018-08-06 12:32:00           2018-08-10 17:18:42   
8389           2018-07-30 12:17:00           2018-08-01 16:08:40   
8424           2018-07-30 12:17:00           2018-07-31 21:22:11   
8425           2018-07-30 12:17:00           2018-07-31 21:22:11   
8448           2018-07-27 17:58:00           2018-08-09 13:44:54   
9778           2018-07-25 13:32:00           2018-08-03 10:03:31   
9940           2018-07-24 15:07:00           2018-07-25 17:57:42   
10297          2018-07-24 09:39:00           2018-07-30 22:54:34   
10505          2018-07-20 21:43:00           2018-07-28 15:26:43   
26600          2018-05-09 11:58:00           2018-05-28 13:51:17   
26838          2018-05-09 11:54:00           2018-05-15 17:05:57   

      order_estimated_delivery_date  order_item_id  \
115                      2018-09-10            1.0   
1868                     2018-09-19            1.0   
3955                     2018-08-30            1.0   
5566                     2018-08-20            1.0   
8389                     2018-08-06            1.0   
8424                     2018-08-06            2.0   
8425                     2018-08-06            1.0   
8448                     2018-08-21            1.0   
9778                     2018-07-31            1.0   
9940                     2018-07-31            1.0   
10297                    2018-08-08            1.0   
10505                    2018-07-30            1.0   
26600                    2018-06-12            1.0   
26838                    2018-06-05            1.0   

                             product_

In [117]:
df23.loc[(df23.product_category_name == 'pc_gamer'), 'product_category_name_english'] = 'pc_gamer'

In [118]:
df23[df23.product_category_name == 'pc_gamer']

order_id                       customer_id  \
1570    4a8493d781a65dfb623103a5dedf44fa  d721d37c2317b409cf35b0de1fd9438b   
8105    1d7542bb5262913fe0516f7943b69a58  a7bf5e6fa311950360efc620e590fa74   
8106    1d7542bb5262913fe0516f7943b69a58  a7bf5e6fa311950360efc620e590fa74   
47977   d7ae3c10c3a4c7a721e64ba401ee745b  8622c629a0890a4ca20d1e7016e1a613   
97208   804e5ddab73def7d5e86b8ae6f27e83b  ae9d8bf13681f31ec761764094e6ab95   
99895   e4aac63ea6f6b75b6d91ead05207fb43  9aa2c4bac83761140a8703ebdf9e49bd   
103893  4821d5af4c2ac98b0f70e47c5d845520  ae1f2f01f136f0172bbceffa5d31613c   
104575  94f94ebc2e26aeea2aa9e8945cb9fb0c  e00a5ec8d72cb3c5ed58f50fb1a55cbf   

       order_status order_purchase_timestamp   order_approved_at  \
1570      delivered      2018-08-18 22:35:15 2018-08-21 04:55:37   
8105      delivered      2018-07-28 23:32:46 2018-07-28 23:45:09   
8106      delivered      2018-07-28 23:32:46 2018-07-28 23:45:09   
47977     delivered      2018-02-20 14:59:40 2018-02-20 15:10:43   
97208     delivered      2017-06-15 14:51:11 2017-06-17 03:43:54   
99895     delivered      2017-05-25 12:51:37 2017-05-25 13:05:42   
103893    delivered      2017-04-24 19:57:30 2017-04-24 20:40:05   
104575    delivered      2017-04-17 14:31:24 2017-04-17 14:45:30   

       order_delivered_carrier_date order_delivered_customer_date  \
1570            2018-08-22 08:25:00           2018-08-29 14:21:54   
8105            2018-07-30 11:33:00           2018-07-31 19:06:21   
8106            2018-07-30 11:33:00           2018-07-31 19:06:21   
47977           2018-02-21 18:27:20           2018-03-03 17:42:36   
97208           2017-06-19 20:32:38           2017-06-30 13:12:08   
99895           2017-05-31 11:35:26           2017-06-02 09:44:29   
103893          2017-04-25 11:08:27           2017-05-02 18:11:33   
104575          2017-04-19 13:53:09           2017-05-04 15:18:33   

       order_estimated_delivery_date  order_item_id  \
1570                      2018-09-11            1.0   
8105                      2018-08-02            2.0   
8106                      2018-08-02            1.0   
47977                     2018-03-19            2.0   
97208                     2017-07-20            1.0   
99895                     2017-06-07            1.0   
103893                    2017-05-15            1.0   
104575                    2017-05-10            1.0   

                              product_id  ... freight_value  \
1570    6727051471a0fc4a0e7737b57bff2549  ...         19.63   
8105    6727051471a0fc4a0e7737b57bff2549  ...          9.69   
8106    6727051471a0fc4a0e7737b57bff2549  ...          9.69   
47977   0105b5323d24fc655f73052694dbbb3a  ...         20.04   
97208   0105b5323d24fc655f73052694dbbb3a  ...         26.19   
99895   0105b5323d24fc655f73052694dbbb3a  ...          8.16   
103893  0105b5323d24fc655f73052694dbbb3a  ...         15.08   
104575  0105b5323d24fc655f73052694dbbb3a  ...         14.67   

       product_category_name  product_name_lenght  product_description_lenght  \
1570                pc_gamer                 60.0                      1532.0   
8105                pc_gamer                 60.0                      1532.0   
8106                pc_gamer                 60.0                      1532.0   
47977               pc_gamer                 59.0                       621.0   
97208               pc_gamer                 59.0                       621.0   
99895               pc_gamer                 59.0                       621.0   
103893              pc_gamer                 59.0                       621.0   
104575              pc_gamer                 59.0                       621.0   

       product_photos_qty  product_weight_g  product_length_cm  \
1570                  3.0             650.0               16.0   
8105                  3.0             650.0               16.0   
8106                  3.0             650.0               16.0   
47977                 4.0           

In [119]:
df23.isnull().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                   0
order_delivered_carrier_date      694
order_delivered_customer_date    1852
order_estimated_delivery_date       0
order_item_id                       0
product_id                          0
seller_id                           0
shipping_limit_date                 0
price                               0
freight_value                       0
product_category_name               0
product_name_lenght                 0
product_description_lenght          0
product_photos_qty                  0
product_weight_g                    1
product_length_cm                   1
product_height_cm                   1
product_width_cm                    1
product_category_name_english       0
dtype: int64

**Insight:**
- Dataframe `df23` sudah dibersihkan dan sudah tidak terdapat missing value pada kolom 'product_category_name_english'.

#### `df4`

Berdasarkan informasi pada tahap assessing data, data yang perlu dibersihkan pada df4 adalah data-data yang mengandung missing value pada kolom review_score. Selain itu kolom pada 'review_comment_title' dan 'review_comment_message' juga akan dihapus. 

In [120]:
# Memperbarui dataframe d4 dengan data df23 yang sudah dibersihkan
df4 = pd.merge(df23, order_reviews_df, on='order_id', how='left')
df4 = df4.sort_values(by='order_purchase_timestamp', ascending=False)
df4.head()

order_id                       customer_id  \
0  54282e97f61c23b78330c15b154c867d  4b7decb9b58e2569548b8b4c8e20e8d7   
1  35a972d7f8436f405b56e36add1a7140  898b7fee99c4e42170ab69ba59be0a8b   
2  03ef5dedbe7492bdae72eec50764c43f  496630b6740bcca28fce9ba50d8a26ef   
3  168626408cb32af0ffaf76711caae1dc  6e353700bc7bcdf6ebc15d6de16d7002   
4  0b223d92c27432930dfe407c6aea3041  e60df9449653a95af4549bbfcb18a6eb   

  order_status order_purchase_timestamp   order_approved_at  \
0      shipped      2018-09-03 09:06:57 2018-09-03 17:40:06   
1    delivered      2018-08-29 15:00:37 2018-08-29 15:10:26   
2    delivered      2018-08-29 14:52:00 2018-08-29 15:05:22   
3    delivered      2018-08-29 14:18:28 2018-08-29 14:30:23   
4    delivered      2018-08-29 14:18:23 2018-08-29 14:31:07   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2018-09-04 15:25:00                           NaT   
1          2018-08-29 16:57:00           2018-08-30 16:23:36   
2          2018-08-29 20:01:00           2018-08-30 16:36:59   
3          2018-08-29 18:51:00           2018-08-30 16:52:31   
4          2018-08-29 15:29:00           2018-08-30 16:24:55   

  order_estimated_delivery_date  order_item_id  \
0                    2018-09-06            1.0   
1                    2018-09-05            1.0   
2                    2018-09-03            1.0   
3                    2018-09-11            1.0   
4                    2018-09-04            1.0   

                         product_id  ... product_length_cm product_height_cm  \
0  b98992ea80b467987a7fbb88e7f2076a  ...              16.0              68.0   
1  d04857e7b4b708ee8b8b9921163edba3  ...              16.0               4.0   
2  c7f27c5bef2338541c772b5776403e6a  ...              28.0               9.0   
3  bdcf6a834e8faa30dac3886c7a58e92e  ...              31.0              16.0   
4  2b4472df15512a2825ae86fd9ae79335  ...              48.0              49.0   

   product_width_cm            product_category_name_english  \
0              32.0  kitchen_dining_laundry_garden_furniture   
1              20.0                    computers_accessories   
2              19.0                           party_supplies   
3              28.0                            health_beauty   
4              49.0  kitchen_dining_laundry_garden_furniture   

                          review_id  review_score      review_comment_title  \
0  6efce984242ca6456d74750810733a03           1.0                       NaN   
1  390c8570648658f80968ae6f77142429           5.0                       NaN   
2  8a8ebb5019795478156fa952acd050b1           3.0                 Recomendo   
3  39eae4e4d255222202c4255e8eee9084           5.0  Chegou rápido e correto    
4  6c50d16eb583d5db7e841b77e89b7045           5.0                       NaN   

   review_comment_message  review_creation_date  review_answer_timestamp  
0              Nao recebi            2018-08-31      2018-08-31 09:51:47  
1                     NaN            2018-08-31      2018-09-09 16:49:27  
2                     NaN            2018-08-31      2018-08-31 20:02:48  
3              Recomendo             2018-08-31      2018-09-02 20:55:45  
4                     NaN            2018-08-31      2018-10-24 16:27:36  

[5 rows x 29 columns]

In [121]:
# Menghapus baris data yang memiliki missing value pada kolom 'review_score'
df4.drop(df4[df4.review_score.isnull()].index, inplace=True)

In [122]:
# Menghapus kolom 'review_comment_title' dan 'review_comment_message'
df4.drop(columns=['review_comment_title', 'review_comment_message'], inplace=True)

In [123]:
# Menghitung jumlah missing value pada setiap kolom di df4
df4.isnull().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                   0
order_delivered_carrier_date      688
order_delivered_customer_date    1772
order_estimated_delivery_date       0
order_item_id                       0
product_id                          0
seller_id                           0
shipping_limit_date                 0
price                               0
freight_value                       0
product_category_name               0
product_name_lenght                 0
product_description_lenght          0
product_photos_qty                  0
product_weight_g                    1
product_length_cm                   1
product_height_cm                   1
product_width_cm                    1
product_category_name_english       0
review_id                           0
review_score                        0
review_creation_date                0
review_answe

**Insight:**
- Dataframe `df4` sudah tidak memiliki missing value pada review_score
- Kolom 'review_comment_title' dan 'review_comment_message' juga sudah dihapus.

#### `df5`

Berdasarkan informasi pada tahap assessing data, data yang perlu dibersihkan pada df5 adalah data-data yang mengandung missing value pada kolom review_score. 

In [124]:
# Memperbarui dataframe d5 dengan data d4 yang sudah dibersihkan
df5 = pd.merge(df4, order_payments_df, on='order_id', how='left')
df5 = df5.sort_values(by='order_purchase_timestamp', ascending=False)
df5.head()

order_id                       customer_id  \
0  54282e97f61c23b78330c15b154c867d  4b7decb9b58e2569548b8b4c8e20e8d7   
1  35a972d7f8436f405b56e36add1a7140  898b7fee99c4e42170ab69ba59be0a8b   
2  03ef5dedbe7492bdae72eec50764c43f  496630b6740bcca28fce9ba50d8a26ef   
3  168626408cb32af0ffaf76711caae1dc  6e353700bc7bcdf6ebc15d6de16d7002   
4  0b223d92c27432930dfe407c6aea3041  e60df9449653a95af4549bbfcb18a6eb   

  order_status order_purchase_timestamp   order_approved_at  \
0      shipped      2018-09-03 09:06:57 2018-09-03 17:40:06   
1    delivered      2018-08-29 15:00:37 2018-08-29 15:10:26   
2    delivered      2018-08-29 14:52:00 2018-08-29 15:05:22   
3    delivered      2018-08-29 14:18:28 2018-08-29 14:30:23   
4    delivered      2018-08-29 14:18:23 2018-08-29 14:31:07   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2018-09-04 15:25:00                           NaT   
1          2018-08-29 16:57:00           2018-08-30 16:23:36   
2          2018-08-29 20:01:00           2018-08-30 16:36:59   
3          2018-08-29 18:51:00           2018-08-30 16:52:31   
4          2018-08-29 15:29:00           2018-08-30 16:24:55   

  order_estimated_delivery_date  order_item_id  \
0                    2018-09-06            1.0   
1                    2018-09-05            1.0   
2                    2018-09-03            1.0   
3                    2018-09-11            1.0   
4                    2018-09-04            1.0   

                         product_id  ... product_width_cm  \
0  b98992ea80b467987a7fbb88e7f2076a  ...             32.0   
1  d04857e7b4b708ee8b8b9921163edba3  ...             20.0   
2  c7f27c5bef2338541c772b5776403e6a  ...             19.0   
3  bdcf6a834e8faa30dac3886c7a58e92e  ...             28.0   
4  2b4472df15512a2825ae86fd9ae79335  ...             49.0   

             product_category_name_english                         review_id  \
0  kitchen_dining_laundry_garden_furniture  6efce984242ca6456d74750810733a03   
1                    computers_accessories  390c8570648658f80968ae6f77142429   
2                           party_supplies  8a8ebb5019795478156fa952acd050b1   
3                            health_beauty  39eae4e4d255222202c4255e8eee9084   
4  kitchen_dining_laundry_garden_furniture  6c50d16eb583d5db7e841b77e89b7045   

   review_score review_creation_date  review_answer_timestamp  \
0           1.0           2018-08-31      2018-08-31 09:51:47   
1           5.0           2018-08-31      2018-09-09 16:49:27   
2           3.0           2018-08-31      2018-08-31 20:02:48   
3           5.0           2018-08-31      2018-09-02 20:55:45   
4           5.0           2018-08-31      2018-10-24 16:27:36   

   payment_sequential  payment_type  payment_installments  payment_value  
0                 1.0       voucher                   1.0         166.46  
1                 1.0   credit_card                   1.0          93.75  
2                 1.0   credit_card                   1.0          33.23  
3                 1.0       voucher                   1.0          61.29  
4                 1.0   credit_card                   8.0         510.96  

[5 rows x 31 columns]

In [125]:
# Menghapus baris data yang memiliki missing value pada kolom review_score
df5.drop(df5[df5.review_score.isnull()].index, inplace=True)

In [126]:
# Menghitung jumlah missing value pada setiap kolom di df5
df5.isnull().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                   0
order_delivered_carrier_date      718
order_delivered_customer_date    1856
order_estimated_delivery_date       0
order_item_id                       0
product_id                          0
seller_id                           0
shipping_limit_date                 0
price                               0
freight_value                       0
product_category_name               0
product_name_lenght                 0
product_description_lenght          0
product_photos_qty                  0
product_weight_g                    1
product_length_cm                   1
product_height_cm                   1
product_width_cm                    1
product_category_name_english       0
review_id                           0
review_score                        0
review_creation_date                0
review_answe

Pada pertanyaan 5, kemungkinan keseluruhan kolom akan digunakan sehingga kolom-kolom yang masih mengandung missing value akan dibersihkan.

In [ ]:
# Menghapus baris data yang memiliki missing value pada df5
df5_clean = df5.dropna()

In [ ]:
# Menghitung jumlah missing value pada df5_clean
df5_clean.isnull().sum()

order_id                         0
customer_id                      0
order_status                     0
order_purchase_timestamp         0
order_approved_at                0
order_delivered_carrier_date     0
order_delivered_customer_date    0
order_estimated_delivery_date    0
order_item_id                    0
product_id                       0
seller_id                        0
shipping_limit_date              0
price                            0
freight_value                    0
product_category_name            0
product_name_lenght              0
product_description_lenght       0
product_photos_qty               0
product_weight_g                 0
product_length_cm                0
product_height_cm                0
product_width_cm                 0
product_category_name_english    0
review_id                        0
review_score                     0
review_creation_date             0
review_answer_timestamp          0
payment_sequential               0
payment_type        

**Insight:**
- Dataframe `df5` sudah tidak memiliki missing value pada kolom 'review_score'
- Dataframe `df5_clean` sudah tidak memiliki missing value pada setiap kolom.

## Exploratory Data Analysis (EDA)

### Explore ...

**Insight:**
- xxx
- xxx

## Visualization & Explanatory Analysis

### Pertanyaan 1:

### Pertanyaan 2:

**Insight:**
- xxx
- xxx

## Analisis Lanjutan (Opsional)

## Conclusion

- Conclution pertanyaan 1
- Conclution pertanyaan 2